### PR TITLE
Jenkins 29977

### PR DESCRIPTION
**Pull request by joseblas**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -205,7 +205,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>apache-httpcomponents-client-4-api</artifactId>
-      <version>4.5.5-3.0</version>
+      <version>4.5.5-2.1</version>
```
Comment (MarkEWaite) [inline pom.xml]: Why is the http components client 4 API version going back to an older version?

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java [hunk 2]
```diff
@@ -977,7 +977,7 @@ public ChangelogCommand changelog() {
         return new ChangelogCommand() {
 
             /** Equivalent to the git-log raw format but using ISO 8601 date format - also prevent to depend on git CLI future changes */
-            public static final String RAW = "commit %H%ntree %T%nparent %P%nauthor %aN <%aE> %ai%ncommitter %cN <%cE> %ci%n%n%w(76,4,4)%s%n%n%b";
+            public static final String RAW = "commit %H%ntree %T%nparent %P%nauthor %aN <%aE> %ai%ncommitter %cN <%cE> %ci%n%n%w(0,4,4)%B%n%n";
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: Can you add an item to the release notes (on the client plugin wiki page) that describes the change of responsibility?

I assume your plan is that the git client plugin will no longer truncate the first line of the commit message.  The git client plugin will deliver the entire first line of the commit message and rely on the git plugin to decide if it should be truncated (current behavior) or not.

Can you confirm that is your plan?

File: src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java [hunk 3]
```diff
@@ -257,6 +257,36 @@ private String randomEmail(String name) {
         return name.replaceAll(" ", ".") + "@middle.earth";
     }
 
+    @Test
+    @Issue("JENKINS-297799")
+    public void testChangelongVeryLong() throws Exception {
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java]:
```suggestion
    /* Changelog first line was truncated on word boundary prior to 
     * 72 characters with git client plugin 2.0+ with CLI git.
     * Was not truncated by git client plugin earlier releases using CLI git.
     * Was not truncated by git client plugin and JGit.
     * Rely on caller to perform truncation of first line if desired.
     * Matching change will be included in git plugin 4.0.0
     * to retain existing truncation behavior.
     */
    public void testChangelogVeryLong() throws Exception {
```

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 2 above)
Comment (MarkEWaite) [reply src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: Replaces #339

File: pom.xml (hunk 1 above)
Comment (joseblas) [reply pom.xml]: Git-plugin needs that, otherwise enforcer will fail the build.

File: src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java (hunk 2 above)
Comment (joseblas) [reply src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java]: Exactly, that is the plan

File: src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java (hunk 3 above)
Comment (joseblas) [reply src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java]: I'd rather go for this:
```
    /**
     *  Changelog was formatted on word boundary prior to
     * 72 characters with git client plugin 2.0+ when using CLI git.
     * Was not truncated by git client plugin using JGit (And Apache version).
     * Rely on caller to truncate first line if desired.
     * Matching change will be included in git plugin 4.0.0
     * to retain existing truncation behavior.
     */

```

Comment (MarkEWaite) [reply src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java]: I believe your proposed comment is incorrect, or at least imprecise when it says:

> Changelog was formatted on word boundary prior to 72 characters with git client plugin 2.0+ when using CLI git.

The **first line** of the changelog was **truncated** on a word boundary prior to 72 characters.  The rest of the changelog was not truncated, only the first line.

File: src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java [hunk 4]
```diff
@@ -257,6 +257,44 @@ private String randomEmail(String name) {
         return name.replaceAll(" ", ".") + "@middle.earth";
     }
 
+    @Test
+    @Issue("JENKINS-297799")
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java]: Please correct the issue number from JENKINS-297799 to JENKINS-29977

File: src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java [hunk 5]
```diff
@@ -257,6 +257,44 @@ private String randomEmail(String name) {
...
+    /**
+     *  Changelog was formatted on word boundary prior to
+     * 72 characters with git client plugin 2.0+ when using CLI git.
+     * Was not truncated by git client plugin using JGit (And Apache version).
+     * Rely on caller to truncate first line if desired.
+     * Matching change will be included in git plugin 4.0.0
+     * to retain existing truncation behavior.
+     */
+    public void testChangelongVeryLong() throws Exception {
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java]: Please correct the method name from testChangelongVeryLong to testChangelogVeryLong

File: pom.xml [hunk 6]
```diff
@@ -43,7 +43,7 @@
 
   <properties>
     <revision>3.0.0-beta6</revision>
-    <changelist>-SNAPSHOT</changelist>
+    <changelist>-rc1791.3a94dfffd1fc</changelist>
```
Comment (MarkEWaite) [inline pom.xml]: Please explain why this change is needed?  What is the benefit of modifying that changelist value?

File: src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java (hunk 3 above)
Comment (joseblas) [reply src/test/java/org/jenkinsci/plugins/gitclient/GitClientTest.java]: The whole message was formatted, not only the first line.
This is the outcome when running the tests (and adding a s.out)


>     Uno Dos Tres Cuatro Cinco Lorem ipsum dolor sit amet, consectetur
>     adipiscing elit. Ut posuere tellus eu efficitur tristique. In iaculis
>     neque in dolor vulputatesollicitudin eget a quam. Donec finibus sapien
>     quis lectus euismod facilisis. Integermassa purus, scelerisque id
>     iaculis ut, blandit vitae velit. Pellentesque lobortisaliquet felis, vel
>     laoreet ipsum tincidunt at. Mauris tellus est, cursus vitae exeget,
>     venenatis auctor eros. Sed sagittis porta odio. Donec ut interdum massa.
>     Aliquamsagittis, mi sit amet sollicitudin elementum, velit quam eleifend
>     nisl, in rhoncusfelis nibh eu nibh. Class aptent taciti sociosqu ad
>     litora torquent per conubia nostra, per inceptos himenaeos. seis
> 
>     asfasfasfasf

whereas with the new formatting is: 

>

>     Uno Dos Tres Cuatro Cinco Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut posuere tellus eu efficitur tristique. In iaculis neque in dolor vulputatesollicitudin eget a quam. Donec finibus sapien quis lectus euismod facilisis. Integermassa purus, scelerisque id iaculis ut, blandit vitae velit. Pellentesque lobortisaliquet felis, vel laoreet ipsum tincidunt at. Mauris tellus est, cursus vitae exeget, venenatis auctor eros. Sed sagittis porta odio. Donec ut interdum massa. Aliquamsagittis, mi sit amet sollicitudin elementum, velit quam eleifend nisl, in rhoncusfelis nibh eu nibh. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos.
>     seis
>     
>     asfasfasfasf

File: pom.xml (hunk 6 above)
Comment (joseblas) [reply pom.xml]: Well, in a previous conversation was decided not to use SNAPSHOTS

Comment (MarkEWaite) [reply pom.xml]: I think you may have misunderstood the earlier conversation.  The git client plugin does not need to have its changelist value modified.  The incrementals process provided by the parent pom delivers artifacts to the incrementals repository automatically with each build of a branch or of a pull request.  

When another plugin needs an unreleased version of a dependency (like the git plugin will need the git client plugin), then the plugin which has the dependency (the git plugin) can refer to the dependency generated by incrementals (the git client plugin build).  

There should be no need to modify the changelist value.
